src/tigerbeetle/main.zig

```diff
@@ -66,37 +66,71 @@ pub fn main() !void {
 
     const allocator = arena.allocator();
 
-    var time_os: TimeOS = .{};
-    const time = time_os.time();
-
     var arg_iterator = try std.process.argsWithAllocator(allocator);
     defer arg_iterator.deinit();
 
     var command = cli.parse_args(&arg_iterator);
 
+    if (command == .version) {
+        try Command.version(allocator, command.version.verbose);
+        return; // Exit early before initializing IO.
+    }
+
+    log_level_runtime = switch (command) {
+        .version => unreachable,
+        .inspect => .info,
+        inline else => |*args| if (args.log_debug) .debug else .info,
+    };
+
+    // Try and init IO early, before a file has even been created, so if it fails (eg, io_uring
+    // is not available) there won't be a dangling file.
+    var io = try IO.init(128, 0);
+    defer io.deinit();
+
+    var time_os: TimeOS = .{};
+    const time = time_os.time();
+
+    var trace_file: ?std.fs.File = null;
+    defer if (trace_file) |file| file.close();
+
+    var statsd_address: ?std.net.Address = null;
+
     switch (command) {
-        .inspect, .version => {},
-        inline else => |*args| {
-            if (args.log_debug) {
-                log_level_runtime = .debug;
+        .start => |*args| {
+            if (args.trace) |path| {
+                trace_file = std.fs.cwd().createFile(path, .{ .exclusive = true }) catch |err| {
+                    log.err("error creating trace file '{s}': {}", .{ path, err });
+                    return err;
+                };
             }
+            if (args.statsd) |address| statsd_address = address;
         },
+        else => {},
     }
 
+    var tracer = try Tracer.init(allocator, time, .unknown, .{
+        .writer = if (trace_file) |file| file.writer().any() else null,
+        .statsd_options = if (statsd_address) |address| .{ .udp = .{
+            .io = &io,
+            .address = address,
+        } } else .log,
+    });
+    defer tracer.deinit(allocator);
+
     switch (command) {
-        .format => |*args| try Command.format(allocator, time, args, .{
+        .format => |*args| try Command.format(allocator, &io, &tracer, args, .{
             .cluster = args.cluster,
```

Comment:
If we are touching a bunch of code here, let's also rename `allocator` to `gpa` in this file!